src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala

```diff
@@ -348,67 +321,29 @@ private class TaskLauncherActor(
-        // In case we don't receive an update a TaskOpRejected message with TASK_OP_REJECTED_TIMEOUT_REASON
-        // reason is scheduled within config.taskOpNotificationTimeout milliseconds. This will trigger another
-        // attempt to launch the task.
-        //
-        // NOTE: this can lead to a race condition where an additional task is launched: in a nutshell if a TaskOp A
-        // is rejected due to an internal timeout, TaskLauncherActor will schedule another TaskOp.Launch B, because
-        // it's under the impression that A has not succeeded/got lost. If the timeout is triggered internally, the
-        // tasksToLaunch count will be increased by 1. The TaskOp A can still be accepted though, and if it is,
-        // two tasks (A and B) might be launched instead of one (given Marathon receives sufficient offers and is
-        // able to create another TaskOp).
-        // This would lead to the app being over-provisioned (e.g. "3 of 2 tasks" launched) but eventually converge
-        // to the target task count when tasks over capacity are killed. With a sufficiently high timeout this case
-        // should be fairly rare.
-        // A better solution would involve an overhaul of the way TaskLaunchActor works and might be
-        // a subject to change in the future.
-        scheduleTaskOpTimeout(instanceOp)
-      }
 
-      OfferMatcherRegistration.manageOfferMatcherStatus()
+    // Mark instance in internal map as provisioned
+    instanceOp.stateOp match {
+      case InstanceUpdateOperation.Provision(instance) =>
+        assert(instanceMap.contains(instance.instanceId), s"Internal task launcher state did not include provisioned instance ${instance.instanceId}")
+        instanceMap += instance.instanceId -> instance
+        logger.info(s"Updated instance map to ${instanceMap.values.map(i => i.instanceId -> i.state.condition)}")
+      case InstanceUpdateOperation.Reserve(instance) =>
+        assert(instanceMap.contains(instance.instanceId), s"Internal task launcher state did not include reserved instance ${instance.instanceId}")
+        instanceMap += instance.instanceId -> instance
+        logger.info(s"Updated instance map to ${instanceMap.values.map(i => i.instanceId -> i.state.condition)}")
+      case other =>
+        logger.info(s"Unexpected updated operation $other")
     }
 
-    updateActorState()
+    OfferMatcherRegistration.manageOfferMatcherStatus()
 
-    logger.debug(s"Request ${instanceOp.getClass.getSimpleName} for instance '${instanceOp.instanceId.idString}', version '${runSpec.version}'. $status")
+    logger.info(s"Request ${instanceOp.getClass.getSimpleName} for instance '${instanceOp.instanceId.idString}', version '${runSpec.version}'. $status")
     promise.trySuccess(MatchedInstanceOps(offer.getId, Seq(InstanceOpWithSource(myselfAsLaunchSource, instanceOp))))
   }
 
-  private[this] def scheduleTaskOpTimeout(instanceOp: InstanceOp): Unit = {
```

Comment:
how come we can drop this timeout?

Reply:
➕ 1

Reply:
I confused it with the delay. Is this is blocker for phase 2? Would it be fine to reintroduce this in phase 3?

Reply:
Please see https://github.com/mesosphere/marathon/pull/6418.